### DISCUSSION TITLE
Meshes with symmetries

### DISCUSSION THREAD

**Discussion by JarosT55** [kinnala/scikit-fem#1097] [Q&A]
**Hi,**

**Skfem** offers predefined meshes with symmetries. Can & how this could be extended to custom meshes?

**scipy.linalg.solve()** offers **assume_a** parameter facilitating dealing with **symmetric matrices**.
Is there a way to trick or build in a **skfem** environment creating a similar mechanism for **scipy.sparse.linalg.spsolve()**?

**Regards
MichaelT**

**Comment by kinnala** [comment]
Symmetric mesh does not lead to a symmetric matrix. However, symmetric bilinear form (u and v can be replaced by one another) leads to a symmetric matrix.

As far as I know, spsolve is a wrapper to SuperLU which has no special support for symmetric matrices. The package scikit-sparse is a wrapper to CHOLMOD which uses the Cholesky decomposition instead of LU decomposition. Using Cholesky decomposition is faster for symmetric and positive definite matrices.

**Reply by JarosT55** [reply]
**Hi Mr. Kinnala,**

_‘Symmetric mesh does not lead to a symmetric matrix.’_

**Yee**, … I realized it by looking at how pre-built symmetrical tri-meshes are initialized.
Essentially, they are flat meshes containing ‘left’ and ‘right’ symmetrical halves.

Well, following such a concept and assembling a symmetrical custom mesh from its one decently sized half (e.g. using: $m_L = m_R.mirrored((0, 1, 0)); m_S = m_L + m_R$ ) would bring no benefit, but a memory and compute time blow-out.

The realistic, potentially beneficial approach (_as I see it_) would be **_‘injecting’ symmetricity at the solver/boundary level_** … but I do not know how to do it … ⚡! The technicality/complexity of the method might be out of reach for **_Skfem Gringo_**, though.

**_Can Skfem offer other tricks from its symmetricity department?_**

**Regards
MichaelT**
